### ISSUE TITLE
Intermittent code-generation failure for C++ => C# wrapper.

### ISSUE THREAD

**Issue by Imisnew2** [swig/swig#1241]
Hi all. We're intermittently noticing an issue with SWIG 3.0.10 where the generated code by SWIG misses some forward type declarations and generates bad code. We're generating a C# wrapper of some C++ methods. We've been trying to figure this out for several months without any success. If anyone here has some insight, it would be greatly appreciated. If it matters, we've only seen this on Ubuntu machines (vs Mac and Windows).

Here's roughly what our SWIG invocation looks like along with our hierarchy of files:

Command Line:
`swig -c++ -csharp -I/path/to/swig/lib -I/path/to/csharp/swig/lib -dllimport MyLib -outdir swig -namespace MyNamespace swig-input.i`

swig-input.i:
```
%module MyModule
%{
#include "fwd-function-decls.hpp"
%}
%include "fwd-class-decls.hpp"
%include "fwd-function-decls.hpp"
```

fwd-class-decls.hpp:
```
#ifndef FWD_CLASS_DECLS_HPP
#define FWD_CLASS_DECLS_HPP
namespace MyNamespace {
enum SomeEnum : unsigned int;
class SomeDataClass;
class SomeOtherClass;
}
#endif // FWD_CLASS_DECLS_HPP
```

fwd-function-decls.hpp:
```
#ifndef FWD_FUNCTION_DECLS_HPP
#define FWD_FUNCTION_DECLS_HPP

#include "fwd-class-decls.hpp"

    #ifdef SWIG
    %pragma(csharp) imclassimports=%{
using System;
using System.Runtime.InteropServices;
    %}
    %typemap(imtype) SWIGTYPE * "System.IntPtr"
    #endif

namespace MyNamespace {
  SomeOtherClass *createSomeOtherClass(SomeDataClass* data, SomeEnum anEnum);
}
#endif // FWD_FUNCTION_DECLS_HPP
```

We expect the following files to be created:

swig-input_wrap.cxx:
```
<snip unrelated SWIG stuff>
// From the %{ } block in swig-input.i 
#include "fwd-function-decls.hpp"

#ifdef __cplusplus
extern "C" {
#endif

SWIGEXPORT void * SWIGSTDCALL CSharp_MyNamespace_createSomeOtherClass___(void * jarg1, int jarg2)
{
  void * jresult ;
  MyNamespace::SomeDataClass *arg1 = (MyNamespace::SomeDataClass *) 0 ;
  MyNamespace::SomeEnum arg2 ;
  MyNamespace::SomeOtherClass *result = 0 ;
  
  arg1 = (MyNamespace::SomeDataClass *)jarg1; 
  arg2 = (MyNamespace::SomeEnum)jarg2;
  result = (MyNamespace::SomeOtherClass *)MyNamespace::createSomeOtherClass(arg1,arg2);
  jresult = (void *)result; 
  return jresult;
}

#ifdef __cplusplus
}
#endif
```

MyModulePINVOKE.cs:
```
namespace MyNamespace {
  class MyModulePINVOKE {
    [global::System.Runtime.InteropServices.DllImport("MyLib", EntryPoint="CSharp_MyNamespace_createSomeOtherClass___")]
    public static extern System.IntPtr createSomeOtherClass(System.IntPtr jarg1, int jarg2);
  }
}
```

However, we notice some weirdness in the resulting generated files every 10-20th time we invoke SWIG. Instead of the above... we get slightly different generated files. Basically, instead of 'MyNamespace::' in the schema-input_wrap.cxx file, there's no namespace qualification on the types, as if SWIG didn't see them. Furthermore, that also affects the MyModulePINVOKE.cs file, `int jarg2` becomes `global::System.Runtime.InteropServices.HandleRef jarg2:

swig-input_wrap.cxx:
```
<snip>
SWIGEXPORT void * SWIGSTDCALL CSharp_MyNamespace_createSomeOtherClass___(void * jarg1, void * jarg2)
{
  void * jresult ;
  SomeDataClass *arg1 = (SomeDataClass *) 0 ;
  SomeEnum arg2 ;
  SomeEnum argp2 ;
  SomeOtherClass *result = 0 ;
  
  arg1 = (SomeDataClass *)jarg1; 
  argp2 = (SomeEnum *)jarg2;
  if (!argp2) {
    SWIG_CSharpSetPendingExceptionArgument(SWIG_CSharpArgumentNullException, "Attempt to dereference null SomeEnum", 0);
    return 0;
  }
  result = (SomeOtherClass *)MyNamespace::createSomeOtherClass(arg1,arg2);
  jresult = (void *)result; 
  return jresult;
}
<snip>
```

MyModulePINVOKE.cs:
```
namespace MyNamespace {
  class MyModulePINVOKE {
    [global::System.Runtime.InteropServices.DllImport("MyLib", EntryPoint="CSharp_MyNamespace_createSomeOtherClass___")]
    public static extern System.IntPtr createSomeOtherClass(System.IntPtr jarg1, global::System.Runtime.InteropServices.HandleRef jarg2);
  }
}
```

**Comment by ojwb** [comment]
3.0.10 was release back in 2016 and there have been quite a lot of C#-related changes since - I'd suggest seeing if you can reproduce this with the latest release (3.0.12 currently) or better still git master.

If it's not already been fixed by changes since 3.0.10, try running swig under valgrind - it sounds like there's probably an uninitialised variable somewhere.

**Comment by Imisnew2** [comment]
Thanks. I'll try running under 3.0.12 for a while and cross my fingers.

**Comment by Imisnew2** [comment]
So, shortly after filing this bug we tracked down another issue that was causing us grief. Basically, we were copying the same file to the same location in multiple 'cp' calls at the same time -- this actually caused 'cp' to fail occasionally, which is how we found the issue. The file being copied? The same fwd-class-decls.hpp in the example above. So, I assume that SWIG has (had?) the same "issue" that our version of 'cp' does (explained better in this link: https://unix.stackexchange.com/questions/116280/cannot-create-regular-file-filename-file-exists), which ended up manifesting as: not complaining a missing file, but not actually reading the file.

Feel free to close this issue report.
